Mod block

l-vincent-l opened the pull request:
# Technical Refactoring

There are lot of things in this PR : 
- First there is a refactorization of read_path: it's now done in a function accepting a visitor.
- Then we remove the journey_pattern_point_connections
- Finally we introduce a next_vj and prev_vj in vehicle_journeys, and we use this in the algorithm.

In the algorithm, at the end of a computation of a journey_pattern if the last taken vehicle_journey has a next_vj, we apply it.
## Benchmark

Centre instance (one with block_id)

|  | mod_block_id | dev |
| :-- | :-: | :-: |
| Memory | 224MB | 928 MB |
| Time per request | ~60ms | ~170ms |

SNCF instance (without block_id)

|  | mod_block_id | dev |
| :-- | :-: | :-: |
| Memory | 275MB | 271MB |
| Time per request | ~150ms | ~140ms |

So when there are no block_id there is an overhead.

l-vincent-l committed:
Kraken: remove journey_pattern_point_connection

l-vincent-l committed:
Kraken: introduce notion of prev, next vehicle journey

This is used to represent how the vehicle journeys are linked
with each other.
This is used when several vj have the same block_id

l-vincent-l committed:
Build_helper: handle prev/next vehicle_journey

l-vincent-l committed:
Ed: link vehicle_journey of the same block_id together

l-vincent-l committed:
Kraken: Use of prev/next vj

We don't use anymore the journey_pattern_connection to handle
block_id.

After computing a journey_pattern, if the used vj has a next_vj
we store it in a queue.

After computing all the journey_patterns, we look at all the vj
in this queue, and we try to apply them on the labels.

l-vincent-l committed:
Kraken: new tests for stay_in

l-vincent-l committed:
Kraken: try to fix stay_in_complexe UT

l-vincent-l committed:
Raptor: add UT of a stay_in without any teleport

l-vincent-l committed:
Builder: better handling of drop_off, pickup allowed

l-vincent-l committed:
Kraken: fix building of raptor_path

l-vincent-l committed:
Kraken: use of read_path in raptor_solutions

l-vincent-l committed:
Kraken: fix apply_vj_extension

And small refactor of raptor_loop

l-vincent-l committed:
Kraken: fix of UT

l-vincent-l committed:
Ed: Update vj with prev and next vj

l-vincent-l committed:
Ed: Fix deletion of vehicle_journeys

antoine-de commented on source/ed/data.cpp:
```diff
@@ -65,10 +62,36 @@ void Data::normalize_uri(){
+        );
+
+    types::VehicleJourney* prev_vj = nullptr;
+    for(auto it=vehicle_journeys.begin(); it!=vehicle_journeys.end(); ++it) {
```
If you don't use the iterator it's better to use the classic for(a:list) form

antoine-de commented on source/ed/data.cpp:
```diff
@@ -65,10 +62,36 @@ void Data::normalize_uri(){
 
+void Data::build_block_id() {
+    /// We want to group vehicle journeys by their block_id
+    /// Two vehicle_journeys vj1 are consecutive if
```
maybe you should update the comment like:
Two vehicle_journeys with the same block_id are consecutive if

l-vincent-l committed:
Kraken: add a not handle yet UT

koch-t commented on source/ed/data.cpp:
```diff
@@ -65,10 +62,36 @@ void Data::normalize_uri(){
+    types::VehicleJourney* prev_vj = nullptr;
+    for(auto it=vehicle_journeys.begin(); it!=vehicle_journeys.end(); ++it) {
+        auto vj = *it;
+        if(prev_vj && prev_vj->block_id != "" && prev_vj->block_id == vj->block_id) {
```
I would add a sanity check to avoid time-travel in blocks.

Furthermore it might be a good idea to set a timelimit for blocktransfers, a transferless connecttion of more than an hour is not likely to be a real transfer. There could be a re-positioning trip in between.

The ambiguity of block_id in GTFS was also cause for debate with the OTP and GTFS community: 
https://groups.google.com/forum/#!searchin/opentripplanner-dev/interlining/opentripplanner-dev/gywq5Z_7Id0/A4J2XZkTPGQJ
https://groups.google.com/forum/#!search/interlining$20gtfs/gtfs-changes/T5QfhkW3Ljw/DsAWK0rWz7QJ

l-vincent-l replied:
To be clear, in your second point, you assume if a transfer is too long it's because the data is wrong?

koch-t replied:
The way block_id is set up leaves much room open for interpretation. Since positioning-trips (trips with empty vehicles) are not included in the GTFS, there is no way of knowing whether passengers are actually allowed to continue in the vehicle the entire block_id.

This would be a perfect valid block in GTFS:
Trip 1 (Station A to B): departure 6:10
Trip 2 (Station B to A): departure 6:40
Trip 3X (Station A to Garage): departure 7:50 (Not for passengers)
Trip 4X (Garage to Station A): departure 16:40 (Not for passengers)
Trip 5 (Station A to B): Departure 17:00

Since trip 3X and 4X are not included, you have no idea that you're not able to sit in the vehicle while transferring from trip 2 to trip 5.

I greatly exaggerated this example but the garage trip could also be a very short circle around the bus-station without passengers. For the latter the solution would be to re-do the way interlining is done in GTFS.
But filtering large interline-dwell's would be good enough to filter this case in the example. Ordinary transfers will ensure a valid route anyway, there just might be a few false negatives missing the notation that passengers can stay on.

antoine-de commented on source/ed/data.cpp:
```diff
@@ -226,6 +249,12 @@ void Data::clean(){
     //Meme chose mais avec les vj
     num_elements = vehicle_journeys.size();
     for(size_t to_erase : erasest) {
+        if(vehicle_journeys[to_erase]->next_vj) {
```
I know it is not directly related to your updates, but I think we should put a general comment for this clean method, and translate the already existing ones.

antoine-de commented on source/ed/ed_persistor.cpp:
```diff
@@ -905,8 +884,32 @@ void EdPersistor::insert_vehicle_journeys(const std::vector<types::VehicleJourne
+        }
+        if(!values.empty()) {
+            if(vj->prev_vj && vj->prev_vj->idx > vehicle_journeys.size()) {
+                std::cout << "pou" << std::endl;
```
you forgot "et", it is important not to misspell 'pouet'

antoine-de commented on source/ed/ed_persistor.cpp:
```diff
@@ -905,8 +884,32 @@ void EdPersistor::insert_vehicle_journeys(const std::vector<types::VehicleJourne
+            query += " WHERE id = ";
+            query += boost::lexical_cast<std::string>(vj->idx);
+            query += ";";
+            LOG4CPLUS_INFO(logger, "query : " << query);
```
you should remove this log too

antoine-de commented on source/ed/types.h:
```diff
@@ -241,9 +227,15 @@ struct VehicleJourney: public Header, Nameable, hasVehicleProperties{
+    VehicleJourney* prev_vj;
+    VehicleJourney* next_vj;
+
+    VehicleJourney(): journey_pattern(NULL), company(NULL), physical_mode(NULL),
```
don't you think it'll be clearer with all the initialization in the member declaration ?

antoine-de replied:
I know it is painfull, but could you translate in the meantime the comments of the class members ?

l-vincent-l replied:
And one hundred balls and a mars with that?

Tristramg replied:
my colleagues are wondering why I’m laughing…

antoine-de commented on source/routing/raptor.cpp:
```diff
@@ -36,37 +36,51 @@ namespace bt = boost::posix_time;
+                                const type::VehicleJourney* prev_vj, type::idx_t boarding_jpp_idx,
+                                DateTime workingDt, const uint16_t l_zone,
+                                const bool disruption_active) {
+    auto& working_labels = labels[count];
```
A comment for the function would make the reading a bit less brain demanding :)

l-vincent-l replied:
Added

antoine-de commented on source/routing/raptor.cpp:
```diff
@@ -341,10 +355,22 @@ struct raptor_visitor {
+    }
+
+    const type::JourneyPatternPoint* get_last_jpp(const type::VehicleJourney* vj) const {
+       return vj->prev_vj->journey_pattern->journey_pattern_point_list.back();
```
is it mandatory for the vj to have a prev_vj ? (do we need to check the pointer ?)

antoine-de commented on source/routing/raptor.h:
```diff
@@ -60,19 +61,18 @@ struct RAPTOR
     ///L'ordre du premier j: public AbstractRouterourney_pattern point de la journey_pattern
     queue_t Q;
+    /// Queue of vehicle journeys used for extension of services
+    std::queue<const type::VehicleJourney*> vj_queue;
```
What is it used for ?

l-vincent-l replied:
It isn't used anymore

antoine-de commented on source/routing/raptor_path.cpp:
```diff
@@ -62,197 +63,120 @@ get_current_stidx_gap(size_t count, type::idx_t journey_pattern_point, const std
 
 
+std::pair<boost::posix_time::ptime, boost::posix_time::ptime>
+handle_st(const type::StopTime* st, DateTime& workingDate, bool clockwise, const type::Data &data){
```
hum, not sure, but I think the side effect on workingDate is not the wanted behaviour so you can skip the '&'.

l-vincent-l replied:
It is wantd

antoine-de commented on source/routing/raptor_path.cpp:
```diff
@@ -62,197 +63,120 @@ get_current_stidx_gap(size_t count, type::idx_t journey_pattern_point, const std
+    boost::posix_time::ptime departure, arrival;
+    if(clockwise) {
+        if(st->is_frequency())
+            DateTimeUtils::update(workingDate, st->f_departure_time(DateTimeUtils::hour(workingDate), !clockwise), !clockwise);
```
hum I don't understand the !clockwise. is it correct to put it everywhere in this function ?
I think comments might have helped me in my understanding :)

antoine-de replied:
Not sure but I think the clockwise and not clockwise cases are the same no ?

antoine-de commented on source/routing/raptor_path.cpp:
```diff
@@ -62,197 +63,120 @@ get_current_stidx_gap(size_t count, type::idx_t journey_pattern_point, const std
-                }
-                item.connection = stop_point_connection;
-            }
+    struct Visitor: public BasePathVisitor {
```
Why do you need a BasePathVisitor ? It is useless isn't it ? you might want to define your visitor elsewhere (and with a more specific name.

antoine-de commented on source/routing/raptor_path_defs.h:
```diff
@@ -0,0 +1,108 @@
+
+namespace navitia { namespace routing {
+template<typename Visitor>
+void read_path(Visitor& v, type::idx_t destination_idx, size_t countb, bool clockwise, bool disruption_active,
```
be merciful, add some comments :)

antoine-de replied:
not sure, but maybe you could split this function in 2 function, one for the connection handling and the other for the vj.

antoine-de commented on source/routing/tests/reverse_raptor_test.cpp:
```diff
@@ -450,6 +451,85 @@ BOOST_AUTO_TEST_CASE(prolongement_service) {
 
+
+
+BOOST_AUTO_TEST_CASE(stay_in_and_one_earlier_with_connection) {
```
wha, nice job for the UT motivation!

l-vincent-l committed:
Ed: Correction of a comment

l-vincent-l committed:
Ed: sanity_check on build_block_id

l-vincent-l committed:
Ed: remove debugging code, and set a log to trace

l-vincent-l committed:
Ed: Improve definition of VehicleJourneys

l-vincent-l committed:
Kraken: Add a comment on apply vj extension

l-vincent-l committed:
Kraken: Add a comment on apply_vj_extension

l-vincent-l committed:
Kraken: Check existence of prev, and next vj pointer

l-vincent-l committed:
Kraken: remove dead code

l-vincent-l committed:
puoeot

l-vincent-l committed:
Kraken: refactoring and commenting of read_path_defs.h

l-vincent-l committed:
Kraken: Add a warning about the side effect of handle_st

l-vincent-l committed:
Kraken: move handle_st to raptor_path_defs and add comments

l-vincent-l committed:
Kraken: fix merge issues

l-vincent-l committed:
Kraken: Update of workingDt was done in the wrong place

This could cause a bad thing in case of overmigdnight

l-vincent-l committed:
Kraken: Better init of labels.

This might lead to something bad and doesn't take that long

l-vincent-l committed:
Kraken: move visitors to a separate file

So it can be inlined

l-vincent-l committed:
Kraken: Add comments on raptor

l-vincent-l committed:
Kraken: move declaration of some variables to a more suitable place

l-vincent-l committed:
Kraken: Change a check so it's more understandable

l-vincent-l committed:
Kraken: Small imrpovement on label marking

We want to update a destination label only if the one of the previous round
of the same jpp wasn't initialized

l-vincent-l committed:
Kraken: Change boarding only if we change the stoptime

l-vincent-l committed:
Kraken: remove useless comment
